File: backend/app/schemas/substation.py
from pydantic import BaseModel
from typing import Optional, List, Dict, Any
from datetime import datetime

class SubstationBase(BaseModel):
    name: str
    code: str
    voltage_level: float
    latitude: float
    longitude: float
    address: Optional[str] = None
    branch_id: int
    description: Optional[str] = None

class SubstationCreate(SubstationBase):
    pass

class SubstationResponse(SubstationBase):
    id: int
    is_active: bool
    created_at: datetime
    updated_at: Optional[datetime]

    class Config:
        from_attributes = True

class ConnectionBase(BaseModel):
    substation_id: int
    power_line_id: int
    connection_type: str
    voltage_level: float
    description: Optional[str] = None

class ConnectionCreate(ConnectionBase):
    pass

class ConnectionResponse(ConnectionBase):
    id: int
    created_at: datetime

    class Config:
        from_attributes = True


# Схемы для VoltageLevel
class VoltageLevelBase(BaseModel):
    name: str
    code: str
    nominal_voltage: float  # кВ
    high_voltage_limit: Optional[float] = None  # кВ
    low_voltage_limit: Optional[float] = None  # кВ
    description: Optional[str] = None

class VoltageLevelCreate(VoltageLevelBase):
    mrid: Optional[str] = None

class VoltageLevelResponse(VoltageLevelBase):
    id: int
    mrid: str
    substation_id: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# Схемы для Bay
class BayBase(BaseModel):
    name: str
    bay_number: str
    bay_type: str  # ввод, отходящая линия, секционный, трансформатор
    description: Optional[str] = None

class BayCreate(BayBase):
    mrid: Optional[str] = None

class BayResponse(BayBase):
    id: int
    mrid: str
    voltage_level_id: int
    created_at: datetime

    class Config:
        from_attributes = True


# Схемы для BusbarSection
class BusbarSectionBase(BaseModel):
    name: str
    section_number: int  # Номер секции (I, II, III)
    nominal_current: Optional[float] = None  # А
    description: Optional[str] = None

class BusbarSectionCreate(BusbarSectionBase):
    mrid: Optional[str] = None

class BusbarSectionResponse(BusbarSectionBase):
    id: int
    mrid: str
    bay_id: int
    created_at: datetime

    class Config:
        from_attributes = True


# Схемы для ConductingEquipment
class ConductingEquipmentBase(BaseModel):
    equipment_type: str  # CIM класс: PowerTransformer, Breaker, Disconnector и т.д.
    name: str
    manufacturer: Optional[str] = None
    model: Optional[str] = None
    serial_number: Optional[str] = None
    specifications: Optional[Dict[str, Any]] = None  # JSON с характеристиками
    installation_date: Optional[datetime] = None
    last_maintenance_date: Optional[datetime] = None
    next_maintenance_date: Optional[datetime] = None
    status: str = "active"  # active, maintenance, decommissioned
    notes: Optional[str] = None

class ConductingEquipmentCreate(ConductingEquipmentBase):
    mrid: Optional[str] = None

class ConductingEquipmentResponse(ConductingEquipmentBase):
    id: int
    mrid: str
    bay_id: int
    created_by: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# Схемы для ProtectionEquipment
class ProtectionEquipmentBase(BaseModel):
    name: str
    protection_type: str  # реле, автомат защиты, УЗО
    manufacturer: Optional[str] = None
    model: Optional[str] = None
    serial_number: Optional[str] = None
    specifications: Optional[Dict[str, Any]] = None  # уставки, ток срабатывания и т.д.
    installation_date: Optional[datetime] = None
    status: str = "active"
    notes: Optional[str] = None

class ProtectionEquipmentCreate(ProtectionEquipmentBase):
    mrid: Optional[str] = None

class ProtectionEquipmentResponse(ProtectionEquipmentBase):
    id: int
    mrid: str
    bay_id: int
    created_by: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# Обновленная схема SubstationResponse с voltage_levels
class SubstationResponseWithStructure(SubstationResponse):
    voltage_levels: List[VoltageLevelResponse] = []

    class Config:
        from_attributes = True

File: frontend/lib/core/models/substation.dart
import 'package:json_annotation/json_annotation.dart';

part 'substation.g.dart';

@JsonSerializable(
  explicitToJson: true,
  includeIfNull: false,
)
class Substation {
  final int id;
  final String mrid;
  final String name;
  @JsonKey(name: 'dispatcher_name')
  final String dispatcherName;
  @JsonKey(name: 'voltage_level')
  final double voltageLevel;
  final double latitude;
  final double longitude;
  final String? address;
  @JsonKey(name: 'region_id')
  final int? regionId;
  @JsonKey(name: 'branch_id')
  final int? branchId;
  @JsonKey(name: 'is_active')
  final bool isActive;
  @JsonKey(name: 'created_at')
  final DateTime createdAt;
  @JsonKey(name: 'updated_at')
  final DateTime? updatedAt;

  const Substation({
    required this.id,
    required this.mrid,
    required this.name,
    required this.dispatcherName,
    required this.voltageLevel,
    required this.latitude,
    required this.longitude,
    this.address,
    this.regionId,
    this.branchId,
    required this.isActive,
    required this.createdAt,
    this.updatedAt,
  });

  factory Substation.fromJson(Map<String, dynamic> json) => _$SubstationFromJson(json);
  Map<String, dynamic> toJson() => _$SubstationToJson(this);
}

@JsonSerializable(
  explicitToJson: true,
  includeIfNull: false,
)
class SubstationCreate {
  final String name;
  @JsonKey(name: 'dispatcher_name')
  final String dispatcherName;
  @JsonKey(name: 'voltage_level')
  final double voltageLevel;
  final double latitude;
  final double longitude;
  final String? address;
  @JsonKey(name: 'branch_id')
  final int? branchId;
  final String? description;

  const SubstationCreate({
    required this.name,
    required this.dispatcherName,
    required this.voltageLevel,
    required this.latitude,
    required this.longitude,
    this.address,
    this.branchId,
    this.description,
  });

  factory SubstationCreate.fromJson(Map<String, dynamic> json) => _$SubstationCreateFromJson(json);
  Map<String, dynamic> toJson() => _$SubstationCreateToJson(this);
}

